gc/default/default.c

```diff
@@ -2832,16 +2835,22 @@ run_final(rb_objspace_t *objspace, VALUE zombie)
     if (FL_TEST_RAW(zombie, FL_FINALIZE)) {
         FL_UNSET(zombie, FL_FINALIZE);
         st_data_t table;
+        unsigned int lev = rb_gc_vm_lock();
```

Comment:
I think we should do the locking in `finalize_list` instead. We lock the VM in `finalize_list` to add the slot back to the freelist, so we're just doing locking and unlocking twice for every zombie object instead of once total, which would be bad for performance.

Reply:
Good catch, I just had to move the lock up by one line.